core/mingw/src/files/FileSystemMingw.kt

```diff
@@ -54,9 +81,66 @@ private const val MAX_PATH_LENGTH = 32767
 
 internal actual fun realpathImpl(path: String): String {
     memScoped {
-        val buffer = allocArray<CHARVar>(MAX_PATH_LENGTH)
-        val len = GetFullPathNameA(path, MAX_PATH_LENGTH.convert(), buffer, null)
+        val buffer = allocArray<WCHARVar>(MAX_PATH_LENGTH)
+        val len = GetFullPathNameW(path, MAX_PATH_LENGTH.convert(), buffer, null)
         if (len == 0u) throw IllegalStateException()
         return buffer.toKString()
     }
 }
+
+internal actual class OpaqueDirEntry(directory: String) : AutoCloseable {
+    private val arena = Arena()
+    private val data = arena.alloc<WIN32_FIND_DATAW>()
+    private var handle: HANDLE? = INVALID_HANDLE_VALUE
+    private var firstName: String? = null
+
+    init {
+        try {
+            val directory0 = if (directory.endsWith('/') || directory.endsWith('\\')) "$directory*" else "$directory/*"
+            handle = FindFirstFileW(directory0, data.ptr)
+            if (handle != INVALID_HANDLE_VALUE) {
+                firstName = data.cFileName.toKString()
+            } else {
+                val le = GetLastError()
+                if (le != ERROR_FILE_NOT_FOUND.toUInt()) {
+                    val strerr = formatWin32ErrorMessage(le)
+                    throw IOException("Can't open directory $directory: $le ($strerr)")
+                }
+            }
+        } catch (th: Throwable) {
+            if (handle != INVALID_HANDLE_VALUE) {
+                CloseHandle(handle)
+            }
+            arena.clear()
+            throw th
+        }
+    }
+
+    actual fun readdir(): String? {
+        if (firstName != null) {
+            return firstName.also { firstName = null }
+        }
+        if (handle == INVALID_HANDLE_VALUE) {
+            return null
+        }
+        if (FindNextFileW(handle, data.ptr) == TRUE) {
+            return data.cFileName.toKString()
+        }
+        val le = GetLastError()
+        if (le == ERROR_NO_MORE_FILES.toUInt()) {
+            return null
+        }
+        val strerr = formatWin32ErrorMessage(le)
+        throw IOException("Can't readdir: $le ($strerr)")
```

Comment:
Does it make sense to include a directory name here too? Like, `"Can't readdir from $directory: ..."`.